subprojects/core/src/integTest/groovy/org/gradle/api/tasks/MissingTaskDependenciesIntegrationTest.groovy

```diff
@@ -435,6 +441,101 @@ class MissingTaskDependenciesIntegrationTest extends AbstractIntegrationSpec imp
         executedAndNotSkipped(":fooReport", ":barReport")
     }
 
+    @Issue("https://github.com/gradle/gradle/issues/20391")
+    def "running tasks in parallel with exclusions does not cause incorrect builds"() {
+        // Configuration caching resolves the inputs once more to store the result in the configuration cache.
+        int numResolvedBeforeTaskExecution = GradleContextualExecuter.configCache ? 2 : 1
+        server.start()
+        file("lib/src/MyClass.java").text = "public class MyClass {}"
+
+        settingsFile """
+            include "dist"
+            include "lib"
+        """
+
+        file("dist/build.gradle").text = """
+            abstract class ZipSrc extends DefaultTask {
+                @Internal
+                int numResolved
+
+                @Internal
+                abstract DirectoryProperty getSources()
+
+                @InputFiles
+                abstract ConfigurableFileCollection getSourceFiles()
+
+                @OutputFile
+                abstract RegularFileProperty getZipFile()
+
+                ZipSrc() {
+                    sourceFiles.from(sources.map {
+                        if (numResolved == ${numResolvedBeforeTaskExecution}) {
+                            println "resolving input"
+                            ${server.callFromBuild("zipFileSnapshottingStarted")}
+                        }
+                        numResolved++
+                        it.asFileTree.matching {
+                            include "src/**"
+                            include "build.gradle"
+                        }
+                    })
```

Comment:
Can we simplify this somewhat? Do we need a separate `sources` property here, or can we inline this part in the configuration of the task?

Reply:
The problem is that this configuration closure needs a way to count access times, that is why I ended up with configuring it in the task so it has access to `countResolved`. I didn't find a way to make the test configuration cache compatible without the extra property.

Reply:
Got it. I think this would be worth mentioning in a comment, too.